File: 20221219_Monday/20221219.go
// Leetcode: 1971. Find if Path Exists in Graph.
// https://leetcode.com/problems/find-if-path-exists-in-graph/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 533 ms, faster than 63.64% of Go online submissions for Find if Path Exists in Graph.
// Memory Usage: 31.5 MB, less than 82.95% of Go online submissions for Find if Path Exists in Graph.
// 2022.12.19 Daily Challenge.
func validPath(n int, edges [][]int, source int, destination int) bool {
  if source == destination {
    return true
  }
  // Adjacency list.
  al := make([][]int,n)
  for i := 0; i < n ; i += 1 {
    al[i] = make([]int,0)
  }
  // Used.
  u := make([]bool,n)
  // Fill Adjacency list.
  for _, v := range edges {
    a,b := v[0], v[1]
    al[a] = append(al[a],b)
    al[b] = append(al[b],a)
  }
  u[source] = true
  q := make([]int,0)
  q = append(q,source)
  for 0 != len(q) {
    x := q[0]
    q = q[1:len(q)]
    for _, y := range al[x] {
      if destination == y { return true}
      if u[y] { continue}
      u[y] = true
      q = append(q,y)
    }
  }
  return false
}

File: 20221219_Monday/20221219.rb
# Leetcode: 1971. Find if Path Exists in Graph.
# https://leetcode.com/problems/find-if-path-exists-in-graph/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 679 ms, faster than 89.47% of Ruby online submissions for Find if Path Exists in Graph.
# Memory Usage: 237.3 MB, less than 89.47% of Ruby online submissions for Find if Path Exists in Graph.
# 2022.12.19 Daily Challenge.
# @param {Integer} n
# @param {Integer[][]} edges
# @param {Integer} source
# @param {Integer} destination
# @return {Boolean}
def valid_path(n, edges, source, destination)
  return true if source == destination
  # Adjacency list.
  al = Array.new(n).map{ [] }
  edges.each do |a,b|
    al[a].push(b)
    al[b].push(a)
  end
  # Used.
  u = Array.new(n,false)
  q = [source] 
  u[source] = true
  while ! q.empty?
    x = q.shift
    al[x].each do |y|
      next if u[y]
      return true if destination == y
      u[y] = true
      q.push(y)
    end
  end
  return false
end
